error.go

```diff
@@ -94,221 +95,199 @@ const (
-	case ErrInvalidPattern:
-		return "value does not match validation pattern"
-	case ErrInvalidRange:
-		return "invalid value range"
-	case ErrInvalidLength:
-		return "invalid value length"
-	}
-	return "unknown error"
-}
-
-// MarshalJSON implements the json marshaler interface.
-func (t *TypedError) MarshalJSON() ([]byte, error) {
-	return json.Marshal(struct {
-		ID    int    `json:"id" xml:"id"`
-		Title string `json:"title" xml:"title"`
-		Msg   string `json:"msg" xml:"msg"`
-	}{
-		ID:    int(t.ID),
-		Title: t.ID.Title(),
-		Msg:   t.Mesg,
-	})
-}
-
-// Error builds an error message from the typed error details.
-func (t *TypedError) Error() string {
-	IncrCounter([]string{"goa", "error", strconv.Itoa(int(t.ID))}, 1.0)
-	js, err := json.Marshal(t)
-	if err != nil {
-		return `{"id":0,"title":"generic","msg":"failed to serialize error"}`
-	}
-	return string(js)
-}
-
-// Error summarizes all the underlying error messages in one JSON array.
-func (m MultiError) Error() string {
-	var buffer bytes.Buffer
-	buffer.WriteString("[")
-	for i, err := range m {
-		txt := err.Error()
-		if _, ok := err.(*TypedError); !ok {
-			b, err := json.Marshal(txt)
-			txt = `{"id":0,"title":"generic","msg":`
-			if err != nil {
-				txt += `"unknown error"}`
-			} else {
-				txt += string(b) + "}"
-			}
-		}
-		buffer.WriteString(txt)
-		if i < len(m)-1 {
-			buffer.WriteString(",")
-		}
-	}
-	buffer.WriteString("]")
-
-	// you can blame rsc for that: https://code.google.com/p/go/issues/detail?id=8592#c3
-	txt := buffer.String()
-	txt = strings.Replace(txt, "\\u0026", "&", -1)
-	txt = strings.Replace(txt, "\\u003c", "<", -1)
-	txt = strings.Replace(txt, "\\u003e", ">", -1)
 
-	return txt
-}
+	// ErrInvalidEncoding is the error produced when a request body fails
+	// to be decoded.
+	ErrInvalidEncoding
 
-// NewBadRequestError wraps the given error into a BadRequestError.
-func NewBadRequestError(err error) *BadRequestError {
-	return &BadRequestError{Actual: err}
-}
+	// ErrInternal is the error produced when unexpected conditions
+	// occur (i.e. bug).
+	ErrInternal
+)
 
-// Error implements error.
-func (b *BadRequestError) Error() string {
-	return b.Actual.Error()
+// CAUTION: classes must be listed in the order as the constants above.
+var classes = []*HTTPErrorClass{
+	{ErrInvalidParamType, "invalid parameter value", 400},
+	{ErrMissingParam, "missing required parameter", 400},
+	{ErrInvalidAttributeType, "invalid attribute type", 400},
+	{ErrMissingAttribute, "missing required attribute", 400},
+	{ErrInvalidEnumValue, "invalid value", 400},
+	{ErrMissingHeader, "missing required HTTP header", 400},
+	{ErrInvalidFormat, "value does not match validation format", 400},
+	{ErrInvalidPattern, "value does not match validation pattern", 400},
+	{ErrInvalidRange, "invalid value range", 400},
+	{ErrInvalidLength, "invalid value length", 400},
+	{ErrInvalidEncoding, "invalid request encoding", 400},
+	{ErrInternal, "internal error", 500},
```

Comment:
could we do something like:

```go
var ErrBrokenTag = NewErrorClass("broken_tag_error", 401)
```

and use it like:
```go
return ErrBrokenTag("this didn't work: %s", "hello")
```

a sort of `fmt.Errorf()` emulation, I'd like that :)